Branch for review lab1 script

YaroslavAggressive opened the pull request:
First iterations of implementation of dfs and bfs visualization with networkx

YaroslavAggressive committed:
Create .keep

YaroslavAggressive committed:
Add script

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,72 @@
+class GraphDrawer:
+
+    def __init__(self, graph: dict, search_path: list):
+        self.graph = nwx.DiGraph(graph)
```
Зачем нам хранить граф в класс GraphDrawer ? может лучше статический метод который отисовывает

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,72 @@
+
+    def build_visual(self, test_num: int, show_in_ide: bool):  # test_id - номер теста, чтобы не путаться
+
+        if not os.path.isdir("test_images//test_graph_" + str(test_num)):  # проверка, не был ли уже создан тест с таким номером
```
магическая константа

JDima replied:
комментарии на английском

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,72 @@
+            os.mkdir("test_images//test_graph_" + str(test_num))
+        else:
+            print("Error: such directory already exists, rename test number")
+            return None  # пока ничего не возвращаем, не был начат процесс визуализации
```
зачем глушить ошибку? в твоем случае у пользователя возникнут вопросы, когда он в ответ получит None, не факт что он будет смотреть консоль

JDima commented on src/GraphVisual.py:
```diff
@@ -0,0 +1,41 @@
+from collections import deque
+
+
+class MyGraph:
```
Лучше граф сделать как датакласс для хранения данных, а методы с обходом графа отделить от графа

JDima commented on src/GraphVisual.py:
```diff
@@ -0,0 +1,41 @@
+        return result
+
+    def BFS(self, start_node):  # в ширину
+        nodes_visited = {node: False for node in self.graph.nodes}
```
не проще множество использовать?

JDima commented on src/GraphVisual.py:
```diff
@@ -0,0 +1,41 @@
+        nodes_stack = [start_node]
+        result = []
+        while nodes_stack:
+            curr_node = nodes_stack.pop(0)
```
Работает за линейное время

JDima commented on src/test_public.py:
```diff
@@ -0,0 +1,110 @@
+        return 'test_{}'.format(self._name)
+
+
+dict_1 = {'A': ['C', 'D'],
```
непонятно по названию что за граф

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,72 @@
+import networkx as nwx
```
нет CI

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,72 @@
+        tmp_num = 0
+        for node in self.search_path:  # рисование dfs и bfs на графике
+            idx = node_names.index(node)
+            color_list[idx]='r'
```
не хватает пробелов

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Added first fixes for lab1 script

YaroslavAggressive committed:
Delete GraphVisual, it was renamed to MyGraph

YaroslavAggressive committed:
Update test_public.py

JDima commented on src/test_public.py:
```diff
@@ -0,0 +1,113 @@
+                'D': ['A', 'C', 'E', 'D', 'D'],
+                'E': ['A']}
+G_6 = nwx.MultiDiGraph(graph_simple)
+res_6 = list(nwx.algorithms.traversal.dfs_tree(G_6, 'A'))
```
очень много повторяющегося кода

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,54 @@
+            nodes_colors[node] = 'gray'
+            result.append(node)
+            for node_tmp in graph.get_neighbours(node):
+                if nodes_colors[node_tmp] == 'white':
```
магическая константа по всему проекту

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,54 @@
+class MyGraph:
+
+    def __init__(self, graph: dict):
+        if isinstance(graph, dict):
```
если не дикт, то мы построим класс и будем работать с ним?

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,54 @@
+from collections import deque
+
+
+class MyGraph:
```
https://docs.python.org/3/library/dataclasses.html

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,91 @@
+    def __init__(self, dir_name: str):
+        self.dir_name = dir_name
+
+    def build_visual(self, graph: nwx.MultiDiGraph, search_path: list, test_num: int, show_in_ide: bool):
```
функция не помещается на экран

JDima replied:
что такое test_num? специально для тестов аргументов?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,91 @@
+import networkx as nwx
+import matplotlib.pyplot as plt
+import imageio
```
в реквесте не вижу файла requirements.txt

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,91 @@
+        else:
+            # print("Error: such directory already exists, rename test number")
+            # return None  # returns nothing yet, it is necessary to show user, that there is an mistake here
+            raise NameError("such directory named <" + TEST_PREFIX + str(test_num) + "> already exists")
```
если просто попробуем сделать os.mkdir(self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(test_num)) не получим ли тоже самое по смыслу?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,91 @@
+
+        # check, that there is no test with such number
+        if not os.path.isdir(self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(test_num)):
+            os.mkdir(self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(test_num))
```
повторяющийся код с условием

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,91 @@
+                                arrowstyle="<|-", style="dashed")
+        if show_in_ide:
+            plt.draw()
+            plt.pause(0.5)  # задержка в отрисовке
```
русские комментарии

YaroslavAggressive committed:
Update requirements.txt

YaroslavAggressive committed:
Updated script with second set of fixes

MyGraph and GraphDrawer reworked using dataclasses, test_public.py was redesigned and eliminated redundant code, and other small fixes

YaroslavAggressive committed:
Update test_public.py

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+@dataclass
+class GraphDrawer:
+
+    dir_name: str = ""
```
непонятно какой смысл от этой переменной

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+class GraphDrawer:
+
+    dir_name: str = ""
+    tests_counter: int = 1
```
не очень хорошо смешивать основную логику с логикой тестов

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+    def build_visual(self, graph: MyGraph, search_path: list, show_in_ide: bool):
+        # check, that there is no test with such number
+        test_path = self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(GraphDrawer.tests_counter)
+        os.mkdir(test_path)
```
если уэе существует директория?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+        images = []
+        for filename in filenames:
+            images.append(imageio.imread(filename))
+        options = {'duration': 2}
```
можно вынести как константу

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,61 @@
+NODE_WORKED = 'black'  # indicator of finish step of dfs/bfs algorithm with current graph node
+
+
+class MyGraph(nwx.MultiDiGraph):
```
к чему приставка My ?

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,61 @@
+        nodes_colors = {node: NODE_NOT_VISITED for node in graph.nodes()}
+        result = []
+
+        def dfs_recursive(node: str):
```
мы точно хотим чтобы этот метод был объявлен внутри функции?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+
+    def build_visual(self, graph: MyGraph, search_path: list, show_in_ide: bool):
+        # check, that there is no test with such number
+        test_path = self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(GraphDrawer.tests_counter)
```
только тестов можно визуализировать?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,89 @@
+
+        list_edges = graph.edges(data=False)
+        node_names = [node for node in graph.nodes]
+        color_list = ['b' for i in range(len(graph.nodes))]
```
магическая константа

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Update requirements.txt

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Upd lab script according to the comments received

Third version of fixes

YaroslavAggressive committed:
Update nested method for dfs algo

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+@dataclass
+class GraphDrawer:
+
+    dir_name: str = ""  # directory name, where user wants to save his visualization result
```
Нам точно нужны эти константы?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+    dir_name: str = ""  # directory name, where user wants to save his visualization result
+    result_idx: int = 0  # index, randomly generated for each function "build_visual" call
+
+    def build_visual(self, graph: Graph, search_path: list, show_in_ide: bool):
```
кажется эти методы можно стедалть статическими

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+        list_edges = graph.edges(data=False)
+        node_names = [node for node in graph.nodes]
+        color_list = [DEFAULT_NODE_COLOR for i in range(len(graph.nodes))]
+        iter_num = 1
```
можно воспользоваться итератором enumerate

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+                 with_labels=True,
+                 arrowstyle="<|-", style="dashed")
+        nwx.draw_networkx_edges(graph, pos, edgelist=list_edges,
+                                node_size=1000,
```
node_size=1000, arrowstyle="<|-", style="dashed" - этот конфиг дублируется - может вынесем в константу?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+                plt.pause(0.5)  # delay in rendering
+
+            # generate path to file
+            filename = self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(self.result_idx) + "//" +\
```
дублируешь код с `create_gif_result`

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,108 @@
+        for filename in filenames:
+            images.append(imageio.imread(filename))
+        # check, that there is no test with such number
+        test_path = self.dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(test_number) + "//" + TEST_PREFIX + str(test_number)
```
почему названия содержат `test` ? это вроде код в основной логике

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,60 @@
+
+
+    #
+    # def __init__(self, graph: dict):
```
нам нужны эти комментарии?

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,60 @@
+
+    @staticmethod
+    def BFS(graph: Graph, start_node):  # breadth search
+        nodes_visited = {node for node in graph.nodes()}
```
лишнее действие, лучше взять пустое множество и в него добавлять

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,60 @@
+                result.append(curr_node)
+                nodes_visited.remove(curr_node)
+            for node in graph.neighbors(curr_node):
+                if node in nodes_visited:
```
судя по названиям ты ходишь по посещенным вершинам

JDima commented on src/test_public.py:
```diff
@@ -0,0 +1,104 @@
+                  'graph with parallel edges bfs test', 'graph with loop bfs test', 'simple graph bfs test']
+
+# tests are calculated on the same set of graphs as dfs
+TEST_CASES_FOR_BFS = [Case(name=str(i) + ": " + test_names_bfs[i], test_num=i + 1, input=graph_dicts[i], start_node=START_POINT,
```
не очень удачно, я поменяю порядок следования в graph_dicts , но здесь не поменяю test_names_bfs  +такое ощущение что test_names_dfs  и test_names_bfs  похожи

YaroslavAggressive committed:
Added  new fixes as per comments (4th iteration)

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,113 @@
+        tmp_path = dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(result_idx)
+        while os.path.isdir(tmp_path):
+            result_idx = randint(INDEX_LEFT_BOUND, INDEX_RIGHT_BOUND)
+            tmp_path = dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(result_idx)
```
дублирующий код со строчками над циклом

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,113 @@
+        result_idx = randint(INDEX_LEFT_BOUND, INDEX_RIGHT_BOUND)
+        tmp_path = dir_name + "//" + TEMP_GRAPH_IMAGE_PREFIX + str(result_idx)
+        while os.path.isdir(tmp_path):
+            result_idx = randint(INDEX_LEFT_BOUND, INDEX_RIGHT_BOUND)
```
не проще взять текущий timestamp и создать папку с таким именем, где хранить результат?

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,113 @@
+class GraphDrawer:
+
+    @staticmethod
+    def build_visual(graph: Graph, search_path: list, dir_name: str, show_in_ide: bool):
```
функция не помещается на экран - может разоббьем её на подфункции?

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,50 @@
+        while nodes_stack:
+            curr_node = nodes_stack.popleft()
+            if curr_node not in nodes_visited:
+                result.append(curr_node)
```
дублируем информацию между `nodes_visited` и `result ` - например, можно было использовать словарь, там сохзраняется порядок добавления

JDima commented on src/test_public.py:
```diff
@@ -0,0 +1,120 @@
+    results.append(tmp_res)
+
+# tests are calculated on the same set of graphs as dfs
+TEST_CASES_FOR_BFS = get_test_cases(graphs_data, START_POINT, results, BFS_NAME)
```
ДЛя структуризации кода - можно разделить код на файлы - данные/bfs тесты/dfs тесты

JDima commented on src/MyGraph.py:
```diff
@@ -0,0 +1,50 @@
+
+    @staticmethod
+    def DFS(graph: Graph, start_node):  # depth search
+        nodes_colors = {node: NODE_NOT_VISITED for node in graph.nodes()}
```
а нам точно нужно сначала все пометить, как не посещенные? если в словаре, значит помечена

JDima commented on src/GraphDrawer.py:
```diff
@@ -0,0 +1,113 @@
+        for filename in filenames:
+            images.append(imageio.imread(filename))
+        # check, that there is no test with such number
+        result_path = GraphDrawer.create_res_path(dir_name, test_number) + "//" + TEST_PREFIX + str(test_number)
```
не очень понятно, почему в основной логике появился TEST_PREFIX

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Rename test_public.py to test_public_dfs.py

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
test_public.py  cut into 3 scripts

YaroslavAggressive committed:
Delete previous test version

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Update requirements.txt

YaroslavAggressive committed:
Delete test_data.py

YaroslavAggressive committed:
Upd test_data

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Rename testing_bfs.py to test_bfs.py

YaroslavAggressive committed:
Rename testing_dfs.py to test_dfs.py

YaroslavAggressive committed:
Update lab_one_test.yml

YaroslavAggressive committed:
Added chenges to visualization

YaroslavAggressive committed:
Added changes in methods
